### ISSUE TITLE
Crash with tiled-linux-64bit-snapshot-2020.05.27

### ISSUE THREAD

**Issue by paaaav** [bjorn/tiled#2809]
This happens when loading a session automatically at startup (restore session in settings enabled). 
The offending line in the session is
```
    "openFiles": [
        "isometric_grass_and_water.tmx"
    ],
```
If changed to
```
    "openFiles": [],
```
there is no crash anymore. Manually loading the map after startup works. Happens also with Tiled snapshot from April.

I've got valgrind dump, do you need something else?
```
$ valgrind ./Tiled 

==238169== Memcheck, a memory error detector
==238169== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==238169== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==238169== Command: ./Tiled
==238169== 
QPainter::begin: Paint device returned engine == 0, type: 1
QPainter::setRenderHint: Painter must be active to set rendering hints
QPainter::setRenderHint: Painter must be active to set rendering hints
QPainter::setWorldTransform: Painter not active
QPainter::worldTransform: Painter not active
QPainter::save: Painter not active
QPainter::save: Painter not active
QPainter::setBrushOrigin: Painter not active
QPainter::restore: Unbalanced save/restore
QPainter::restore: Unbalanced save/restore
QPainter::setWorldTransform: Painter not active
QPainter::save: Painter not active
QPainter::setOpacity: Painter not active
==238169== Invalid read of size 8
==238169==    at 0x5B3325F: ??? (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libtiled.so)
==238169==    by 0x5B1B730: Tiled::IsometricRenderer::drawTileLayer(QPainter*, Tiled::TileLayer const*, QRectF const&) const (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libtiled.so)
==238169==    by 0x56EAD7F: QGraphicsScenePrivate::draw(QGraphicsItem*, QPainter*, QTransform const*, QTransform const*, QRegion*, QWidget*, double, QTransform const*, bool, bool) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EBB07: QGraphicsScenePrivate::drawSubtreeRecursive(QGraphicsItem*, QPainter*, QTransform const*, QRegion*, QWidget*, double, QTransform const*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EAE6E: QGraphicsScenePrivate::draw(QGraphicsItem*, QPainter*, QTransform const*, QTransform const*, QRegion*, QWidget*, double, QTransform const*, bool, bool) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EBB07: QGraphicsScenePrivate::drawSubtreeRecursive(QGraphicsItem*, QPainter*, QTransform const*, QRegion*, QWidget*, double, QTransform const*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EAE6E: QGraphicsScenePrivate::draw(QGraphicsItem*, QPainter*, QTransform const*, QTransform const*, QRegion*, QWidget*, double, QTransform const*, bool, bool) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EBB07: QGraphicsScenePrivate::drawSubtreeRecursive(QGraphicsItem*, QPainter*, QTransform const*, QRegion*, QWidget*, double, QTransform const*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EC0C9: QGraphicsScenePrivate::drawItems(QPainter*, QTransform const*, QRegion*, QWidget*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x570E649: QGraphicsView::paintEvent(QPaintEvent*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x543378F: QWidget::event(QEvent*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x54D817D: QFrame::event(QEvent*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==238169== 
==238169== 
==238169== Process terminating with default action of signal 11 (SIGSEGV)
==238169==  Access not within mapped region at address 0x0
==238169==    at 0x5B3325F: ??? (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libtiled.so)
==238169==    by 0x5B1B730: Tiled::IsometricRenderer::drawTileLayer(QPainter*, Tiled::TileLayer const*, QRectF const&) const (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libtiled.so)
==238169==    by 0x56EAD7F: QGraphicsScenePrivate::draw(QGraphicsItem*, QPainter*, QTransform const*, QTransform const*, QRegion*, QWidget*, double, QTransform const*, bool, bool) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EBB07: QGraphicsScenePrivate::drawSubtreeRecursive(QGraphicsItem*, QPainter*, QTransform const*, QRegion*, QWidget*, double, QTransform const*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EAE6E: QGraphicsScenePrivate::draw(QGraphicsItem*, QPainter*, QTransform const*, QTransform const*, QRegion*, QWidget*, double, QTransform const*, bool, bool) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EBB07: QGraphicsScenePrivate::drawSubtreeRecursive(QGraphicsItem*, QPainter*, QTransform const*, QRegion*, QWidget*, double, QTransform const*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EAE6E: QGraphicsScenePrivate::draw(QGraphicsItem*, QPainter*, QTransform const*, QTransform const*, QRegion*, QWidget*, double, QTransform const*, bool, bool) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EBB07: QGraphicsScenePrivate::drawSubtreeRecursive(QGraphicsItem*, QPainter*, QTransform const*, QRegion*, QWidget*, double, QTransform const*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x56EC0C9: QGraphicsScenePrivate::drawItems(QPainter*, QTransform const*, QRegion*, QWidget*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x570E649: QGraphicsView::paintEvent(QPaintEvent*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x543378F: QWidget::event(QEvent*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==    by 0x54D817D: QFrame::event(QEvent*) (in /home/pav/bin/Tiled/tiled-linux-64bit-snapshot-2020.05.27/usr/lib/libQt5Widgets.so.5)
==238169==  If you believe this happened as a result of a stack
==238169==  overflow in your program's main thread (unlikely but
==238169==  possible), you can try to increase the size of the
==238169==  main thread stack using the --main-stacksize= flag.
==238169==  The main thread stack size used in this run was 8388608.
==238169== 
==238169== HEAP SUMMARY:
==238169==     in use at exit: 22,988,068 bytes in 115,887 blocks
==238169==   total heap usage: 560,355 allocs, 444,468 frees, 129,082,069 bytes allocated
==238169== 
==238169== LEAK SUMMARY:
==238169==    definitely lost: 6,824 bytes in 28 blocks
==238169==    indirectly lost: 34,999 bytes in 103 blocks
==238169==      possibly lost: 45,151 bytes in 481 blocks
==238169==    still reachable: 22,901,094 bytes in 115,275 blocks
==238169==                       of which reachable via heuristic:
==238169==                         newarray           : 11,512 bytes in 39 blocks
==238169==                         multipleinheritance: 35,472 bytes in 47 blocks
==238169==         suppressed: 0 bytes in 0 blocks
==238169== Rerun with --leak-check=full to see details of leaked memory
==238169== 
==238169== For lists of detected and suppressed errors, rerun with: -s
==238169== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 18 from 4)
Segmentation fault (core dumped)
```

**Comment by bjorn** [comment]
Hmm, that is very interesting that there is a `QPaintEvent` but the painter does not have a valid paint engine. Still, the invalid read appears to happen in `libtiled.so`, though that could be via a Qt header file that doesn't show up on the stack trace. It's a pity we don't have line numbers.

Since it's paint engine related, did you by any chance enable hardware accelerated rendering? If so, please try if it does work when launching Tiled with the `--disable-opengl` parameter.

Another thing that could be helpful is when you can compile Tiled yourself and run that build in gdb/valgrind, so that we could get line numbers and potentially more information about the context of the crash.

**Comment by paaaav** [comment]
I think it was enabled, yes. I tried --disable-opengl and it did not crash. Then re-enabled hardware accelerated rendering in the settings and it still does not crash anymore. I don't know why and how. Can this be a graphics driver issue? I remember it happened constantly, weird. I will try to reproduce again. If I can reproduce it, I will compile Tiled myself as you have suggested.

Another thing: I noticed right now that Tiled writes the session to the path [...]/Arbeitsfl�che (invalid encoding)/[...] instead of [...]/Arbeitsfläche/[...]. It seems it handles umlauts and maybe other non-ASCII-characters in the session save path wrongly ($ls says the directory name is 'Arbeitsfl'$'\344''che') . It did work properly in snapshot-2020.04.10. Tiled recreates the original directory structure (only folders, all empty but **the** session file).
